### ISSUE TITLE
Want to enable Pre-Shared Key communication using SSL Service.

### ISSUE THREAD

**Issue by anil291987** [apple/swift-nio-ssl#230]
Hi,
 I am connecting with  my server with certificate and PSK communication. Able to connect with certificate with server. But not able to figure out how to do [PSK](https://en.wikipedia.org/wiki/Pre-shared_key) communication using [swift-nio-ssl.](https://github.com/apple/swift-nio-ssl) If you have required more information from my side. Please let me know.

**Comment by Lukasa** [comment]
Yes, we should support this. Pre-Shared Keys are complex enough that we probably want to do this with a delegate protocol for users to implement. The trickiest part of this will be arranging for the PSKs to be able to secure stored on the heap. For that use-case, we probably need to bring over [`SecureBytes`](https://github.com/apple/swift-crypto/blob/master/Sources/Crypto/Util/SecureBytes.swift) from `swift-crypto`.

The delegate would be something like:

```swift
public protocol NIOSSLPreSharedKeyCache {
    func preSharedKey(identity: [UInt8]) -> SecureBytes
}
```

Once we have that, wiring up the rest of the interface should be easy enough, I think.

However, this is not a terribly high priority feature for the NIO core team, so we are unlikely to rush to implement this feature. It's on our roadmap, we'll keep this issue open, but if anyone else feels like making the contribution they should go for it.

**Comment by Lukasa** [comment]
As a breadcrumb, if you're only running your application on Apple platforms, [SwiftNIO Transport Services](https://github.com/apple/swift-nio-transport-services) gives you access to Network.framework with SwiftNIO, which already has PSK support.

**Comment by Lukasa** [comment]
Resolved using #369.
